### PR TITLE
Support named types in Type and TypeSignature

### PR DESCRIPTION
**Pull request by rongrong**
* Named type's type signature will be formatted as
  qualified_type_name:base_type_signature.
* Introduces TypeWithName, which delegate all execution aspects of the
  type to its base type.
* Make SingatureBinder, type coercion, type constraint, etc to work with
  TypeWithName.

Test plan - (Please fill in how you tested your changes)

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureBase.java [hunk 1]
```diff
@@ -39,33 +43,58 @@ public static TypeSignatureBase of(QualifiedObjectName name)
...
+        checkArgument(typeName != null, "typeName is null");
+        checkArgument(validateName(typeName.getObjectName()), "Bad characters in typeName: %s", typeName);
         this.standardTypeBase = Optional.empty();
-        this.typeBase = Optional.of(typeBase);
+        this.typeName = Optional.of(typeName);
+    }
+
+    private TypeSignatureBase(QualifiedObjectName typeName, String standardTypeBase)
+    {
+        checkArgument(typeName != null && standardTypeBase != null, "typeName or standardTypeBase is null");
```
Comment (kaikalur) [inline presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureBase.java:71]: It can only be a standard type?

Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureBase.java:71]: The "standard type" here refers to all types defined in `StandardTypes.java` which includes all presto types.

**Review by arhimondr — APPROVED** [review]
Quick summary of what has been discussed offline.

Originally we thought about making the type name aliases only a frontend concept. This entails two main problems

1. Connectors currently return a physical type as part of the column definition. If the named type is only a frontend concept it will be not possible to propagate a named type through presto core (think of `CREATE TABLE LIKE ...`)
2. Error messages could be very obscure if the semantic name is lost along the way to the execution (think of a type USER that is physically represented as a row with 30+ fields, the end user still want's to see `USER` as a type of a column in the error message)

Ideally it would be nice to separate these two concepts. Rename `Type` to something like `StorageType` or `PhysicalType`. And have a separate entity, called `NamedType` or `SemanticType` that includes two fields, `storageType` and `semanticName` and gradually do the refactor across the code. This way it will be easy to understand what kind of type we are dealing with at each level, and do the gradual refactoring. However the main concern with this approach is that it is time consuming.

To save developing time it was decided not to add another concept, and extend the `Type` interface instead, adding a field that will store the semantic name for a type. This approach is less time consuming, as it requires significantly less effort put in refactoring. However it obscures in what places what type is used, and may introduce bugs that will not be discovered compile time (e.g.: instanceof checks in the physical readers / writers code).

While I'm still in favor of the first approach (gradual refactor), given the complexity I'm okay with moving forward with the current approach. 

@wenleix @highker @shixuan-fan I would really appreciate to know your opinion on this.

**Comment by rongrong** [comment]
> Quick summary of what has been discussed offline.
> 
> Originally we thought about making the type name aliases only a frontend concept. This entails two main problems
> 
> 1. Connectors currently return a physical type as part of the column definition. If the named type is only a frontend concept it will be not possible to propagate a named type through presto core (think of `CREATE TABLE LIKE ...`)
> 2. Error messages could be very obscure if the semantic name is lost along the way to the execution (think of a type USER that is physically represented as a row with 30+ fields, the end user still want's to see `USER` as a type of a column in the error message)
> 
> Ideally it would be nice to separate these two concepts. Rename `Type` to something like `StorageType` or `PhysicalType`. And have a separate entity, called `NamedType` or `SemanticType` that includes two fields, `storageType` and `semanticName` and gradually do the refactor across the code. This way it will be easy to understand what kind of type we are dealing with at each level, and do the gradual refactoring. However the main concern with this approach is that it is time consuming.
> 
> To save developing time it was decided not to add another concept, and extend the `Type` interface instead, adding a field that will store the semantic name for a type. This approach is less time consuming, as it requires significantly less effort put in refactoring. However it obscures in what places what type is used, and may introduce bugs that will not be discovered compile time (e.g.: instanceof checks in the physical readers / writers code).
> 
> While I'm still in favor of the first approach (gradual refactor), given the complexity I'm okay with moving forward with the current approach.
> 
> @wenleix @highker @shixuan-fan I would really appreciate to know your opinion on this.

We also discussed about refactoring the code after this first approach of extending `Type` interface to add name approach, which we gradually replace `Type` with `TypeWithName`(semantic type) in the frontend to eventually remove the necessity of having `TypeWithName` extending `Type`. The difference between doing the refactoring first vs. doing it later is primarily how fast we can make this feature available, and the ability to do refactoring incrementally or not.

File: presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedType.java [hunk 2]
```diff
@@ -37,4 +37,9 @@ public TypeSignature getPhysicalTypeSignature()
     {
         return representation;
     }
+
+    public TypeSignature toTypeSignature()
```
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedType.java:41]: You mean remove this function and directly use `new TypeSignature(this)` when this function is used?

**Pull request merged**